tsl/src/compression/compression.c

```diff
@@ -2215,3 +2215,129 @@ decompress_batches_for_insert(ChunkInsertState *cis, Chunk *chunk, TupleTableSlo
 
 	table_close(in_rel, NoLock);
 }
+
+static ScanKeyData *
+build_scankeys_for_update_delete(int32 hypertable_id, RowDecompressor decompressor, List *attname,
+								 List *attval, List *strategy)
+{
+	int index = 0;
+	ScanKeyData *scankeys = NULL;
+	TypeCacheEntry *tce;
+	ListCell *name;
+	ListCell *val;
+	ListCell *strat;
+
+	if (attname->length)
+	{
+		scankeys = palloc0(attname->length * 1 * sizeof(ScanKeyData));
+		forthree (name, attname, val, attval, strat, strategy)
+		{
+			Node *col_name = lfirst(name);
+			Const *col_attval = lfirst(val);
+			int op_strategy = lfirst_int(strat);
+			FormData_hypertable_compression *fd =
+				ts_hypertable_compression_get_by_pkey(hypertable_id, strVal(col_name));
+			if (COMPRESSIONCOL_IS_SEGMENT_BY(fd))
+			{
+				int attno = 0;
+				Form_pg_attribute attr = TupleDescAttr(decompressor.in_desc, attno);
+				Datum value;
+
+				value = col_attval->constvalue;
+				for (attno = 1; attno <= decompressor.in_desc->natts; attno++)
+				{
+					attr = TupleDescAttr(decompressor.in_desc, attno - 1);
+					if (namestrcmp(&(attr->attname), strVal(col_name)) == 0)
+						break;
+				}
+				tce = lookup_type_cache(attr->atttypid, TYPECACHE_EQ_OPR_FINFO);
+				if (!col_attval->constisnull)
+				{
+					ScanKeyEntryInitialize(&scankeys[index],
+										   0, /* flags */
+										   attno,
+										   op_strategy,
+										   InvalidOid, /* No strategy subtype. */
+										   attr->attcollation,
+										   tce->eq_opr_finfo.fn_oid,
+										   value);
+				}
+				index++;
+			}
+		}
+	}
+	return scankeys;
+}
+
+/*
+ * This method will:
+ *  1.scan compressed chunk
+ *  2.decompress the row
+ *  3.insert decompressed rows to uncompressed chunk
+ *  4.delete this row from compressed chunk
+ */
+bool
+decompress_batches_for_update_delete(Chunk *ch, List *attname, List *attval, List *strategy,
+									 bool *chunk_status_changed)
+{
+	Relation chunk_rel;
+	Relation comp_chunk_rel;
+	Chunk *comp_chunk;
+	HeapTuple compressed_tuple;
+	RowDecompressor decompressor;
+	Snapshot snapshot;
+	int num_scankeys = 0;
+	ScanKeyData *scankeys = NULL;
+
+	snapshot = GetTransactionSnapshot();
+	chunk_rel = table_open(ch->table_id, RowExclusiveLock);
+	comp_chunk = ts_chunk_get_by_id(ch->fd.compressed_chunk_id, true);
+	comp_chunk_rel = relation_open(comp_chunk->table_id, RowExclusiveLock);
+
+	decompressor = build_decompressor(comp_chunk_rel, chunk_rel);
+
+	if (attname && attname->length)
+	{
+		scankeys = build_scankeys_for_update_delete(ch->fd.hypertable_id,
+													decompressor,
+													attname,
+													attval,
+													strategy);
+		num_scankeys = attname->length;
+	}
+
+	TableScanDesc heapScan = table_beginscan(comp_chunk_rel, snapshot, num_scankeys, scankeys);
```

Comment:
because, as Sven had pointed out, SK_SEARCHNULL and SK_SEARCHNOTNULL are not supported by heapscan, I think some logic has to be added for handling NULL values in segmentby columns. For example: 
```
create table exam (time timestamptz, a int, b int);
 select create_hypertable('exam', 'time');

insert INTO exam values (now(), 1, 1);
insert INTO exam values (now(), 2, 2);

insert INTO exam values (now(), NULL, 2);
insert INTO exam values (now(), NULL, 2);

ALTER TABLE exam set (timescaledb.compress, timescaledb.compress_segmentby='a');

select compress_chunk(show_chunks('exam'));

update exam set b = 4 where a = NULL;
```
In this example currently the rows where a = NULL will not get updated.

Reply:
oh wait that was just stupid of me, sorry for the confusion... 😅 this should be `a IS NULL` and that does update the respective rows, so you can ignore this!